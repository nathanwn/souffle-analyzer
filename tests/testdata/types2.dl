.type Exp = Const { val: number }
          | Var { id: symbol}
          | Add { e1: Exp, e2: Exp }

.decl has_expr(exp: Exp)

has_expr($Const(1)).
has_expr($Var("x")).
has_expr($Add($Const(2), $Const(3))).

.decl produces_value(val: number)

produces_value(val) :- has_expr($Const(val)).
produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.

.output produces_value
