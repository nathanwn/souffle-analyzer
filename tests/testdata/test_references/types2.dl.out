types2.dl
 0 | .type Exp = Const { val: number }
 1 |           | Var { id: symbol}
 2 |           | Add { e1: Exp, e2: Exp }
 3 |
 4 | .decl has_expr(exp: Exp)
 5 |
 6 | has_expr($Const(1)).
 7 | has_expr($Var("x")).
 8 | has_expr($Add($Const(2), $Const(3))).
 9 |
10 | .decl produces_value(val: number)
11 |
12 | produces_value(val) :- has_expr($Const(val)).
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
14 |
15 | .output produces_value

=====

-- Cursor range --
types2.dl
 0 | .type Exp = Const { val: number }
   |       ~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                                ~~~

types2.dl
 4 | .decl has_expr(exp: Exp)
   |                     ~~~

_____

-- Cursor range --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
 1 |           | Var { id: symbol}
   |             ~~~
-- References --
types2.dl
 1 |           | Var { id: symbol}
   |             ~~~

types2.dl
 7 | has_expr($Var("x")).
   |           ~~~

_____

-- Cursor range --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |             ~~~
-- References --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |             ~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |           ~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                  ~~~

_____

-- Cursor range --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                       ~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                                ~~~

types2.dl
 4 | .decl has_expr(exp: Exp)
   |                     ~~~

_____

-- Cursor range --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                                ~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                                ~~~

types2.dl
 4 | .decl has_expr(exp: Exp)
   |                     ~~~

_____

-- Cursor range --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |                     ~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                       ~~~

types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |                                ~~~

types2.dl
 4 | .decl has_expr(exp: Exp)
   |                     ~~~

_____

-- Cursor range --
types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
 7 | has_expr($Var("x")).
   |           ~~~
-- References --
types2.dl
 1 |           | Var { id: symbol}
   |             ~~~

types2.dl
 7 | has_expr($Var("x")).
   |           ~~~

_____

-- Cursor range --
types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |           ~~~
-- References --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |             ~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |           ~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                  ~~~

_____

-- Cursor range --
types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
10 | .decl produces_value(val: number)
   |       ~~~~~~~~~~~~~~
-- References --
types2.dl
10 | .decl produces_value(val: number)
   |       ~~~~~~~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   | ~~~~~~~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   | ~~~~~~~~~~~~~~

types2.dl
15 | .output produces_value
   |         ~~~~~~~~~~~~~~

_____

-- Cursor range --
types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   | ~~~~~~~~~~~~~~
-- References --
types2.dl
10 | .decl produces_value(val: number)
   |       ~~~~~~~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   | ~~~~~~~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   | ~~~~~~~~~~~~~~

types2.dl
15 | .output produces_value
   |         ~~~~~~~~~~~~~~

_____

-- Cursor range --
types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   | ~~~~~~~~~~~~~~
-- References --
types2.dl
10 | .decl produces_value(val: number)
   |       ~~~~~~~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   | ~~~~~~~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   | ~~~~~~~~~~~~~~

types2.dl
15 | .output produces_value
   |         ~~~~~~~~~~~~~~

_____

-- Cursor range --
types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~
-- References --
types2.dl
 4 | .decl has_expr(exp: Exp)
   |       ~~~~~~~~

types2.dl
 6 | has_expr($Const(1)).
   | ~~~~~~~~

types2.dl
 7 | has_expr($Var("x")).
   | ~~~~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   | ~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                        ~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                        ~~~~~~~~

_____

-- Cursor range --
types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                  ~~~
-- References --
types2.dl
 2 |           | Add { e1: Exp, e2: Exp }
   |             ~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |           ~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                  ~~~

_____

-- Cursor range --
types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~
-- References --
types2.dl
 0 | .type Exp = Const { val: number }
   |             ~~~~~

types2.dl
 6 | has_expr($Const(1)).
   |           ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                ~~~~~

types2.dl
 8 | has_expr($Add($Const(2), $Const(3))).
   |                           ~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   |                                  ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                       ~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   |                                                   ~~~~~

_____

-- Cursor range --
types2.dl
15 | .output produces_value
   |         ~~~~~~~~~~~~~~
-- References --
types2.dl
10 | .decl produces_value(val: number)
   |       ~~~~~~~~~~~~~~

types2.dl
12 | produces_value(val) :- has_expr($Const(val)).
   | ~~~~~~~~~~~~~~

types2.dl
13 | produces_value(val) :- has_expr($Add($Const(c1), $Const(c2))), val = c1 + c2.
   | ~~~~~~~~~~~~~~

types2.dl
15 | .output produces_value
   |         ~~~~~~~~~~~~~~

_____
